addons/l10n_in/models/product_template.py

```diff
@@ -2,21 +2,28 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo import api, models, fields, _
-from odoo.exceptions import ValidationError
 
 
 class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
     l10n_in_hsn_code = fields.Char(string="HSN/SAC Code", help="Harmonized System Nomenclature/Services Accounting Code")
     l10n_in_hsn_description = fields.Char(string="HSN/SAC Description", help="HSN/SAC description is required if HSN/SAC code is not provided.")
+    l10n_in_hsn_warning = fields.Text(string="HSC/SAC warning", compute="_compute_l10n_in_hsn_warning")
 
-    @api.constrains('l10n_in_hsn_code')
-    def _check_hsn_code_validation(self):
+    @api.depends('sale_ok', 'l10n_in_hsn_code')
+    def _compute_l10n_in_hsn_warning(self):
+        digit_suffixes = {
+                '4': _("either 4, 6 or 8"),
+                '6': _("either 6 or 8"),
+                '8': _("8")
+            }
+        active_hsn_code_digit_len = self.env.companies.mapped('l10n_in_hsn_code_digit')
+        # Remove False value if any company have False value in l10n_in_hsn_code_digit
+        active_hsn_code_digit_len = list(filter(bool, active_hsn_code_digit_len))
         for record in self:
-            company = record.company_id or self.env.company
-            minimum_hsn_len = company.l10n_in_hsn_code_digit
-            check_hsn = record.l10n_in_hsn_code and minimum_hsn_len
-            if check_hsn and len(record.l10n_in_hsn_code) < int(minimum_hsn_len):
-                error_message = _("As per your HSN/SAC code validation, minimum %s digits HSN/SAC code is required.", minimum_hsn_len)
-                raise ValidationError(error_message)
+            check_hsn = record.sale_ok and record.l10n_in_hsn_code and active_hsn_code_digit_len
+            if check_hsn and (not record.l10n_in_hsn_code.isnumeric() or len(record.l10n_in_hsn_code) not in [4, 6, 8] or len(record.l10n_in_hsn_code) < int(max(active_hsn_code_digit_len))):
+                record.l10n_in_hsn_warning = _("HSN code field must consist solely of digits and be %s in length.", digit_suffixes.get(max(active_hsn_code_digit_len)))
```

Comment:
This is not translatable in general: we would need the whole sentence to be sure that we can translated in all languages.
Does it work at least in India?
You are missing the `.po` diff btw.

Reply:
Please don't write long lines
```suggestion
                record.l10n_in_hsn_warning = _(
                    "HSN code field must consist solely of digits and be %s in length.",
                    digit_suffixes.get(max(active_hsn_code_digit_len))
                )
```